src/contracts/facilitators/gsm/GsmL2.sol

```diff
@@ -0,0 +1,601 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.10;
+
+import {VersionedInitializable} from '@aave/core-v3/contracts/protocol/libraries/aave-upgradeability/VersionedInitializable.sol';
+import {IERC20} from '@aave/core-v3/contracts/dependencies/openzeppelin/contracts/IERC20.sol';
+import {GPv2SafeERC20} from '@aave/core-v3/contracts/dependencies/gnosis/contracts/GPv2SafeERC20.sol';
+import {EIP712} from '@openzeppelin/contracts/utils/cryptography/EIP712.sol';
+import {SignatureChecker} from '@openzeppelin/contracts/utils/cryptography/SignatureChecker.sol';
+import {SafeCast} from '@openzeppelin/contracts/utils/math/SafeCast.sol';
+import {AccessControl} from '@openzeppelin/contracts/access/AccessControl.sol';
+
+import {Errors} from 'src/contracts/facilitators/gsm/libraries/Errors.sol';
+import {IGhoFacilitator} from 'src/contracts/gho/interfaces/IGhoFacilitator.sol';
+import {IGhoToken} from 'src/contracts/gho/interfaces/IGhoToken.sol';
+import {IGsmPriceStrategy} from 'src/contracts/facilitators/gsm/priceStrategy/interfaces/IGsmPriceStrategy.sol';
+import {IGsmFeeStrategy} from 'src/contracts/facilitators/gsm/feeStrategy/interfaces/IGsmFeeStrategy.sol';
+import {IGsm} from 'src/contracts/facilitators/gsm/interfaces/IGsm.sol';
+import {IGsmL2} from 'src/contracts/facilitators/gsm/interfaces/IGsmL2.sol';
+
+/**
+ * @title GsmL2
+ * @author Aave
+ * @notice GHO Stability Module. It provides buy/sell facilities to go to/from an underlying asset to/from GHO.
+ * @dev To be covered by a proxy contract.
+ */
+contract GsmL2 is IGsm, IGsmL2, AccessControl, VersionedInitializable, EIP712 {
+  using GPv2SafeERC20 for IERC20;
+  using SafeCast for uint256;
+
+  /// @inheritdoc IGsm
+  bytes32 public constant CONFIGURATOR_ROLE = keccak256('CONFIGURATOR_ROLE');
+
+  /// @inheritdoc IGsm
+  bytes32 public constant TOKEN_RESCUER_ROLE = keccak256('TOKEN_RESCUER_ROLE');
+
+  /// @inheritdoc IGsm
+  bytes32 public constant SWAP_FREEZER_ROLE = keccak256('SWAP_FREEZER_ROLE');
+
+  /// @inheritdoc IGsm
+  bytes32 public constant LIQUIDATOR_ROLE = keccak256('LIQUIDATOR_ROLE');
+
+  /// @inheritdoc IGsm
+  bytes32 public constant BUY_ASSET_WITH_SIG_TYPEHASH =
+    keccak256(
+      'BuyAssetWithSig(address originator,uint256 minAmount,address receiver,uint256 nonce,uint256 deadline)'
+    );
+
+  /// @inheritdoc IGsm
+  bytes32 public constant SELL_ASSET_WITH_SIG_TYPEHASH =
+    keccak256(
+      'SellAssetWithSig(address originator,uint256 maxAmount,address receiver,uint256 nonce,uint256 deadline)'
+    );
+
+  /// @inheritdoc IGsm
+  address public immutable GHO_TOKEN;
+
+  /// @inheritdoc IGsm
+  address public immutable UNDERLYING_ASSET;
+
+  /// @inheritdoc IGsm
+  address public immutable PRICE_STRATEGY;
+
+  /// @inheritdoc IGsm
+  mapping(address => uint256) public nonces;
+
+  address internal _ghoTreasury;
+  address internal _feeStrategy;
+  address internal _liquidityProvider;
+  uint128 internal _exposureCap;
+  uint128 internal _ghoLiquidity;
```

Comment:
no getter for this. Also, it's impossible to know how much was used and the total amount of gho liquidity that was provided to the contract